### PR TITLE
Improve agent shutdown handling

### PR DESCRIPTION
**Pull request by jodh-intel**
Fixed logic used to handle static agent tracing.

For a standard (untraced) hypervisor shutdown, the runtime kills the VM
process once the workload has finished. But if static agent tracing is
enabled, the agent running inside the VM is responsible for the
shutdown. The existing code handled this scenario but did not wait for
the hypervisor process to end. The outcome of this being that the
console watcher thread was killed too early.

Although not a problem for an untraced system, if static agent tracing
was enabled, the logs from the hypervisor would be truncated, missing the
crucial final stages of the agents shutdown sequence.

The fix necessitated adding a new parameter to the `stopSandbox()` API,
which if true requests the runtime hypervisor logic simply to wait for
the hypervisor process to exit rather than killing it.

Fixes: #1696.

PR also includes some code refactoring along with a potential speed improvement for hypervisor shutdown.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

**Comment by jodh-intel** [comment]
I've marked this as not needing to be backported. However, we may want to discuss this.

File: src/runtime/virtcontainers/utils/utils.go [hunk 1]
```diff
@@ -312,3 +315,75 @@ func ConvertNetlinkFamily(netlinkFamily int32) pbTypes.IPFamily {
...
+// WaitLocalProcess waits for the specified process for up to timeoutSecs seconds.
+//
+// Notes:
+//
+// - If the initial signal is zero, the specified process is assumed to be
+//   attempting to stop itself.
+// - If the initial signal is not zero, it will be sent to the process before
+//   checking if it is running.
+// - If the process has not ended after the timeout value, it will be forcibly killed.
+func WaitLocalProcess(pid int, timeoutSecs uint, initialSignal syscall.Signal, logger *logrus.Entry) error {
```
Comment (jodh-intel) [inline src/runtime/virtcontainers/utils/utils.go:328]: ftr, this name was chosen to denote that it refers to a local (host-side) process and is unrelated to the `WaitProcess()` sandbox interface.

File: src/runtime/virtcontainers/utils/utils.go [hunk 2]
```diff
@@ -312,3 +315,75 @@ func ConvertNetlinkFamily(netlinkFamily int32) pbTypes.IPFamily {
...
+			pidRunning = false
+			break
+		}
+
+		if err = syscall.Kill(pid, syscall.Signal(0)); err != nil {
+			pidRunning = false
+			break
+		}
+
+		if time.Since(startTime).Seconds() >= float64(timeoutSecs) {
```
Comment (devimc) [inline src/runtime/virtcontainers/utils/utils.go:369]: uhmmm I'm afraid :worried: - this looks prone to hacks, what happen if system's time is the reset to `startTime`?  maybe we shouldn't rely on the system's time .. ?  can you use a channel timeout ?

**Review by devimc — COMMENTED** [review]
thanks @jodh-intel - I've left some comments and questions

File: src/runtime/virtcontainers/utils/utils.go (hunk 2 above)
Comment (jodh-intel) [reply src/runtime/virtcontainers/utils/utils.go:369]: Hi @devimc - this is simply a rework of what we currently do... in three different places ;) Maybe a follow-on PR could be used to address improving the logic.

File: src/runtime/virtcontainers/acrn.go [hunk 3]
```diff
@@ -511,36 +511,16 @@ func (a *Acrn) stopSandbox(ctx context.Context) (err error) {
...
-			break
-		}
-
-		// Let's avoid to run a too busy loop
-		time.Sleep(time.Duration(50) * time.Millisecond)
+	if waitOnly {
+		// NOP
+		shutdownSignal = syscall.Signal(0)
+	} else {
+		shutdownSignal = syscall.SIGINT
```
[the remark this answers is not in the document]
Comment (jodh-intel) [reply src/runtime/virtcontainers/acrn.go]: Fixed (here and below). Oddly, I started out with that logic ;)

File: src/runtime/virtcontainers/utils/utils.go (hunk 2 above)
Comment (jodh-intel) [reply src/runtime/virtcontainers/utils/utils.go:369]: I guess I could keep the top commit (to show the code was just moved), and then add another commit to rework using a channel...

Comment (devimc) [reply src/runtime/virtcontainers/utils/utils.go:369]: :+1: ok, sounds good to me, thanks

Comment (jodh-intel) [reply src/runtime/virtcontainers/utils/utils.go]: @devimc - branch updated.

**Review by devimc — APPROVED** [review]
thanks @jodh-intel

**Review by fidencio — COMMENTED** [review]
@jodh-intel, it looks good to me, but I don't feel comfortable approving it and having it merged before @cmaf goes through this series.

Thannks for the contribution!

**Comment by jodh-intel** [comment]
Hi @cmaf - could you tal please?

**Review by cmaf — APPROVED** [review]
thanks @jodh-intel

**Pull request merged**